### ISSUE TITLE
Build issue on Solaris 11, due to HAVE_HTONLL not being defined

### ISSUE THREAD

**Issue by gmarler** [net-amqp-rabbitmq/net-amqp-rabbitmq#66]
When I try to build this module on Solaris 11, the following results:

...
/opt/solarisstudio12.4/bin/cc -c    -D_REENTRANT -features=conststrings -xtarget=generic -mr -xO3 -I/usr/gnu/include -m64 -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DPERL_USE_SAFE_PUTENV -I/perfwork/.cpanm/work/1443622827.18674/Net-AMQP-RabbitMQ-0.900000 -features=conststrings -xtarget=generic -mr    -DVERSION=\"0.900000\"  -DXS_VERSION=\"0.900000\" -KPIC "-I/usr/perl5/5.22.0/lib/sun4-solaris-thread-multi-64/CORE"  -DAMQ_PLATFORM='"Perl_solaris_v5.22.0"' RabbitMQ.c
"amqp.h", line 1043: warning: syntax error:  empty declaration
"amqp.h", line 2038: warning: syntax error:  empty declaration
"amqp_private.h", line 325: syntax error before or at: val
"amqp_private.h", line 325: warning: old-style declaration or incorrect type for: val
"amqp_private.h", line 325: warning: old-style declaration or incorrect type for: u
"amqp_private.h", line 325: warning: old-style declaration or incorrect type for: u
"amqp_private.h", line 325: syntax error before or at: .
"amqp_private.h", line 325: warning: old-style declaration or incorrect type for: whole
"amqp_private.h", line 325: non-constant initializer: op "NAME"
"amqp_private.h", line 325: warning: old-style declaration or incorrect type for: u
"amqp_private.h", line 325: syntax error before or at: .
"amqp_private.h", line 325: zero or negative subscript
"amqp_private.h", line 325: warning: old-style declaration or incorrect type for: halves
"amqp_private.h", line 325: undefined struct/union member: halves
"amqp_private.h", line 325: warning: left operand of "." must be struct/union object
"amqp_private.h", line 325: cannot dereference non-pointer type
"amqp_private.h", line 325: warning: {}-enclosed initializer required
"amqp_private.h", line 325: too many array initializers
"amqp_private.h", line 325: warning: old-style declaration or incorrect type for: u
"amqp_private.h", line 325: syntax error before or at: .
"amqp_private.h", line 325: warning: old-style declaration or incorrect type for: halves
"amqp_private.h", line 325: identifier redefined: halves
        current : array[1] of int
        previous: array[-2] of int : "amqp_private.h", line 325
"amqp_private.h", line 325: warning: left operand of "." must be struct/union object
"amqp_private.h", line 325: cannot dereference non-pointer type
"amqp_private.h", line 325: warning: {}-enclosed initializer required
"amqp_private.h", line 325: non-constant initializer: op "U*"
"amqp_private.h", line 325: syntax error before or at: return
"amqp_private.h", line 325: warning: old-style declaration or incorrect type for: u
"amqp_private.h", line 325: warning: old-style declaration or incorrect type for: whole
"amqp_private.h", line 326: syntax error before or at: val
"amqp_private.h", line 326: warning: old-style declaration or incorrect type for: val
"amqp_private.h", line 326: warning: old-style declaration or incorrect type for: u
"amqp_private.h", line 326: warning: old-style declaration or incorrect type for: u
"amqp_private.h", line 326: syntax error before or at: .
"amqp_private.h", line 326: warning: old-style declaration or incorrect type for: whole
"amqp_private.h", line 326: identifier redefined: whole
        current : int
        previous: int : "amqp_private.h", line 325
"amqp_private.h", line 326: non-constant initializer: op "NAME"
"amqp_private.h", line 326: warning: old-style declaration or incorrect type for: u
"amqp_private.h", line 326: syntax error before or at: .
"amqp_private.h", line 326: zero or negative subscript
"amqp_private.h", line 326: warning: old-style declaration or incorrect type for: halves
"amqp_private.h", line 326: identifier redefined: halves
        current : array[-2] of int
        previous: array[1] of int : "amqp_private.h", line 325
"amqp_private.h", line 326: warning: left operand of "." must be struct/union object
"amqp_private.h", line 326: cannot dereference non-pointer type
"amqp_private.h", line 326: warning: {}-enclosed initializer required
"amqp_private.h", line 326: too many array initializers
"amqp_private.h", line 326: warning: old-style declaration or incorrect type for: u
"amqp_private.h", line 326: syntax error before or at: .
"amqp_private.h", line 326: warning: old-style declaration or incorrect type for: halves
"amqp_private.h", line 326: identifier redefined: halves
        current : array[1] of int
        previous: array[-2] of int : "amqp_private.h", line 326
"amqp_private.h", line 326: warning: left operand of "." must be struct/union object
"amqp_private.h", line 326: cannot dereference non-pointer type
"amqp_private.h", line 326: warning: {}-enclosed initializer required
"amqp_private.h", line 326: non-constant initializer: op "U*"
"amqp_private.h", line 326: syntax error before or at: return
"amqp_private.h", line 326: warning: old-style declaration or incorrect type for: u
"amqp_private.h", line 326: warning: old-style declaration or incorrect type for: whole
"amqp_private.h", line 329: type specifier can not be used as array size expression qualifier
"amqp_private.h", line 329: warning: no explicit type given
"amqp_private.h", line 329: type specifier can not be used as array size expression qualifier
"amqp_private.h", line 329: warning: no explicit type given
"amqp_private.h", line 329: type specifier can not be used as array size expression qualifier
"amqp_private.h", line 329: warning: no explicit type given
"amqp_private.h", line 330: type specifier can not be used as array size expression qualifier
"amqp_private.h", line 330: warning: no explicit type given
"amqp_private.h", line 330: type specifier can not be used as array size expression qualifier
"amqp_private.h", line 330: warning: no explicit type given
"amqp_private.h", line 330: type specifier can not be used as array size expression qualifier
"amqp_private.h", line 330: warning: no explicit type given
"amqp_private.h", line 331: type specifier can not be used as array size expression qualifier
"amqp_private.h", line 331: warning: no explicit type given
"amqp_private.h", line 331: type specifier can not be used as array size expression qualifier
"amqp_private.h", line 331: warning: no explicit type given
"amqp_private.h", line 331: type specifier can not be used as array size expression qualifier
"amqp_private.h", line 331: warning: no explicit type given
"amqp_private.h", line 332: type specifier can not be used as array size expression qualifier
"amqp_private.h", line 332: warning: no explicit type given
"amqp_private.h", line 332: type specifier can not be used as array size expression qualifier
"amqp_private.h", line 332: warning: no explicit type given
"amqp_private.h", line 332: type specifier can not be used as array size expression qualifier
"amqp_private.h", line 332: warning: no explicit type given
"amqp_private.h", line 334: type specifier can not be used as array size expression qualifier
"amqp_private.h", line 334: warning: no explicit type given
"amqp_private.h", line 346: type specifier can not be used as array size expression qualifier
"amqp_private.h", line 346: warning: no explicit type given
"amqp_private.h", line 360: type specifier can not be used as array size expression qualifier
"amqp_private.h", line 361: warning: no explicit type given
"amqp_private.h", line 361: warning: old-style declaration or incorrect type for: amqp_abort
"perl_math_int64.h", line 17: type specifier can not be used as array size expression qualifier
"perl_math_int64.h", line 17: warning: no explicit type given
"perl_math_int64.h", line 17: warning: old-style declaration or incorrect type for: perl_math_int64_load
"perl_math_int64.h", line 25: wrong storage class keyword in array size expression
"perl_math_int64.h", line 26: wrong storage class keyword in array size expression
"perl_math_int64.h", line 26: type specifier can not be used as array size expression qualifier
"perl_math_int64.h", line 26: warning: no explicit type given
"perl_math_int64.h", line 26: warning: old-style declaration or incorrect type for: math_int64_c_api_min_version
"perl_math_int64.h", line 27: wrong storage class keyword in array size expression
"perl_math_int64.h", line 27: type specifier can not be used as array size expression qualifier
"perl_math_int64.h", line 27: warning: no explicit type given
"perl_math_int64.h", line 27: warning: old-style declaration or incorrect type for: math_int64_c_api_max_version
"perl_math_int64.h", line 34: wrong storage class keyword in array size expression
"perl_math_int64.h", line 36: wrong storage class keyword in array size expression
"perl_math_int64.h", line 36: type specifier can not be used as array size expression qualifier
"perl_math_int64.h", line 36: warning: no explicit type given
"perl_math_int64.h", line 36: warning: old-style declaration or incorrect type for: math_int64_c_api_SvI64OK
"perl_math_int64.h", line 38: wrong storage class keyword in array size expression
"perl_math_int64.h", line 40: wrong storage class keyword in array size expression
"perl_math_int64.h", line 40: type specifier can not be used as array size expression qualifier
"perl_math_int64.h", line 40: warning: no explicit type given
"perl_math_int64.h", line 40: warning: old-style declaration or incorrect type for: math_int64_c_api_SvU64OK
"perl_math_int64.h", line 42: wrong storage class keyword in array size expression
"perl_math_int64.h", line 44: wrong storage class keyword in array size expression
"perl_math_int64.h", line 46: wrong storage class keyword in array size expression
"perl_math_int64.h", line 67: warning: newline not last character in file
"RabbitMQ.xs", line 22: wrong storage class keyword in array size expression
"RabbitMQ.xs", line 34: type specifier can not be used as array size expression qualifier
"RabbitMQ.xs", line 34: warning: no explicit type given
"RabbitMQ.xs", line 34: warning: old-style declaration or incorrect type for: hash_to_amqp_table
"RabbitMQ.xs", line 35: type specifier can not be used as array size expression qualifier
"RabbitMQ.xs", line 35: warning: no explicit type given
"RabbitMQ.xs", line 35: warning: old-style declaration or incorrect type for: array_to_amqp_array
"RabbitMQ.xs", line 39: type specifier can not be used as array size expression qualifier
"RabbitMQ.xs", line 39: warning: no explicit type given
"RabbitMQ.xs", line 39: warning: old-style declaration or incorrect type for: die_on_error
"RabbitMQ.xs", line 51: type specifier can not be used as array size expression qualifier
"RabbitMQ.xs", line 51: warning: no explicit type given
"RabbitMQ.xs", line 51: warning: old-style declaration or incorrect type for: die_on_amqp_error
"RabbitMQ.xs", line 54: warning: function expects to return value: die_on_amqp_error
"RabbitMQ.xs", line 200: type specifier can not be used as array size expression qualifier
"RabbitMQ.xs", line 200: warning: no explicit type given
"RabbitMQ.xs", line 200: warning: old-style declaration or incorrect type for: internal_recv
"RabbitMQ.xs", line 206: type specifier can not be used as array size expression qualifier
"RabbitMQ.xs", line 206: warning: no explicit type given
"RabbitMQ.xs", line 206: warning: old-style declaration or incorrect type for: result
"RabbitMQ.xs", line 207: type specifier can not be used as array size expression qualifier
"RabbitMQ.xs", line 207: warning: no explicit type given
"RabbitMQ.xs", line 207: warning: old-style declaration or incorrect type for: is_utf8_body
"RabbitMQ.xs", line 209: type specifier can not be used as array size expression qualifier
"RabbitMQ.xs", line 209: warning: no explicit type given
"RabbitMQ.xs", line 209: warning: old-style declaration or incorrect type for: i
"RabbitMQ.xs", line 214: type specifier can not be used as array size expression qualifier
"RabbitMQ.xs", line 214: warning: no explicit type given
"RabbitMQ.xs", line 214: cannot recover from previous errors
cc: acomp failed for RabbitMQ.c
*** Error code 2
make: Fatal error: Command failed for target `RabbitMQ.o'
...

Turns out that all that is simply due to HAVE_HTONLL not being defined, when it should be for Solaris.

This one line change fixes everything, just as confirmation of the above assertion:

--- amqp_private.h.orig Wed Sep 30 10:23:50 2015
+++ amqp_private.h      Wed Sep 30 10:31:12 2015
@@ -321,6 +321,7 @@
 # error Endianness not known
 #endif
 
+#define HAVE_HTONLL
 #ifndef HAVE_HTONLL
 DECLARE_XTOXLL(hton)
 DECLARE_XTOXLL(ntoh)

So the question is, what needs to be done so that HAVE_HTONLL is properly defined?

**Comment by markwellis** [comment]
so this is the same as #57? i aim to get that fixed soonish

in the meantime, does this work around help?

https://github.com/net-amqp-rabbitmq/net-amqp-rabbitmq/issues/57#issuecomment-139582792

**Comment by manchicken** [comment]
Mark, isn't that the same define which was put in for OSX recently?

**Comment by markwellis** [comment]
@manchicken it is, but I've not thought of a nice way to have it defined on the platforms where it's needed, so it's not fixed yet.

I'll have more of a think on how we can do it without building the librabbitmq-c library properly, which we can't as it needs cmake and that's an extra dependency I don't want to force on people. That it should be pretty easy to hardcode a list of OS's that need -DHAVE_HTONLL and just add to it when we find more that need it.

The problem with that is I don't have a Solaris or Mac box to test on!

**Comment by markwellis** [comment]
it seems that librabbitmq-c tests if htonll exists with the following (if it compiles it adds -DHAVE_HTONLL). now how to make that work for us i don't know!

```
#include <arpa/inet.h>

int main() {
    return htonll(0);
}
```

**Comment by y** [comment]
Use Devel::CheckLib to test for it. Haven't tested this, just glanced at the docs:

    use Devel::CheckLib;
    my $need_htonl = check_lib(
        function => 'return htonll(0);
        header   => 'arpa/inet.h',
    );

**Comment by markwellis** [comment]
yes! that should work. thanks @y don't know why i didn't think of this since it's in another project i maintain!

**Comment by manchicken** [comment]
I have a Mac box.

**Comment by manchicken** [comment]
Okay, so I think that I may have resolved this in PR#73. There still seem
to be some travis.yml issues, I'm trying to resolve those. I may not get to
them tonight.

**Comment by markwellis** [comment]
this is released in version 1.1, which will be on cpan shortly
